test/integration/dnspolicy_controller_test.go

```diff
@@ -1220,5 +1220,162 @@ var _ = Describe("DNSPolicy", Ordered, func() {
+						SetIdentifier: "s07c46.lb-" + lbHash + ".test.example.com",
+						RecordTTL:     60,
+						Labels:        nil,
+						ProviderSpecific: v1alpha1.ProviderSpecific{
+							{
+								Name:  "weight",
+								Value: "120",
+							},
+						},
+					},
+					{
+						DNSName: "lb-" + lbHash + ".test.example.com",
+						Targets: []string{
+							"default.lb-" + lbHash + ".test.example.com",
+						},
+						RecordType:    "CNAME",
+						SetIdentifier: "default",
+						RecordTTL:     300,
+						ProviderSpecific: v1alpha1.ProviderSpecific{
+							{
+								Name:  "geo-code",
+								Value: "*",
+							},
+						},
+					},
+					{
+						DNSName: "test.example.com",
+						Targets: []string{
+							"lb-" + lbHash + ".test.example.com",
+						},
+						RecordType:    "CNAME",
+						SetIdentifier: "",
+						RecordTTL:     300,
+					},
+				}
+
+				err := k8sClient.Get(ctx, client.ObjectKey{Name: gateway.Name, Namespace: gateway.Namespace}, gateway)
+				Expect(err).NotTo(HaveOccurred())
+				patch := client.MergeFrom(gateway.DeepCopy())
+				addressType := mgcgateway.MultiClusterIPAddressType
+				gateway.Status.Addresses = []gatewayv1beta1.GatewayAddress{
+					{
+						Type:  &addressType,
+						Value: fmt.Sprintf("%s/%s", "kind-mgc-control-plane", TestAttachedRouteAddressOne),
+					},
+					{
+						Type:  &addressType,
+						Value: fmt.Sprintf("%s/%s", "kind-mgc-control-plane", TestAttachedRouteAddressTwo),
+					},
+				}
+				Expect(k8sClient.Status().Patch(ctx, gateway, patch)).To(BeNil())
+
+				err = k8sClient.Get(ctx, client.ObjectKey{Name: gateway.Name, Namespace: gateway.Namespace}, gateway)
+				Expect(err).NotTo(HaveOccurred())
+				Expect(gateway.Spec.Listeners).NotTo(BeNil())
+				// add another listener, should result in 4 probes
+				typedHostname := gatewayv1beta1.Hostname(OtherAttachedRouteName)
+				otherListener := gatewayv1beta1.Listener{
+					Name:     gatewayv1beta1.SectionName(OtherAttachedRouteName),
+					Hostname: &typedHostname,
+					Port:     gatewayv1beta1.PortNumber(80),
+					Protocol: gatewayv1beta1.HTTPProtocolType,
+				}
+
+				gateway.Spec.Listeners = append(gateway.Spec.Listeners, otherListener)
+				Expect(k8sClient.Update(ctx, gateway)).To(BeNil())
+
+				probeList := &v1alpha1.DNSHealthCheckProbeList{}
+				Eventually(func() error {
+					Expect(k8sClient.List(ctx, probeList, &client.ListOptions{Namespace: testNamespace})).To(BeNil())
+					if len(probeList.Items) != 4 {
+						return fmt.Errorf("expected %v probes, got %v", 4, len(probeList.Items))
+					}
+					return nil
+				}, TestTimeoutLong, TestRetryIntervalMedium).Should(BeNil())
+				Expect(len(probeList.Items)).To(Equal(4))
+
+				//
+				Eventually(func() error {
+					getProbe := &v1alpha1.DNSHealthCheckProbe{}
+					if err = k8sClient.Get(ctx, client.ObjectKey{Name: fmt.Sprintf("%s-test-dns-policy-%s", TestAttachedRouteAddressOne, OtherAttachedRouteName), Namespace: testNamespace}, getProbe); err != nil {
```

Comment:
Noting that the probe being set to unhealthy here relates to the other listener/dnsrecord such that the expected endpoints are a full set as this probe doesn't belong to the listener/dnsrecord that's tested at the end of this test.